File: client/cpp/src/saga_context.cpp
#include "angzarr/saga_context.hpp"

#include <iomanip>
#include <sstream>

namespace angzarr {

SagaContext::SagaContext(const std::vector<EventBook>& destination_books) {
    for (const auto& book : destination_books) {
        add_destination(book);
    }
}

void SagaContext::add_destination(const EventBook& book) {
    if (book.has_cover() && !book.cover().domain().empty()) {
        std::string root_bytes;
        if (book.cover().has_root()) {
            root_bytes = book.cover().root().value();
        }
        std::string key = make_key(book.cover().domain(), root_bytes);
        destinations_[key] = book;
    }
}

uint32_t SagaContext::get_sequence(const std::string& domain,
                                   const std::string& aggregate_root) const {
    std::string key = make_key(domain, aggregate_root);
    auto it = destinations_.find(key);
    if (it == destinations_.end() || it->second.pages_size() == 0) {
        return 1;
    }
    const auto& last_page = it->second.pages(it->second.pages_size() - 1);
    return last_page.sequence() + 1;
}

const EventBook* SagaContext::get_destination(const std::string& domain,
                                              const std::string& aggregate_root) const {
    std::string key = make_key(domain, aggregate_root);
    auto it = destinations_.find(key);
    return it != destinations_.end() ? &it->second : nullptr;
}

bool SagaContext::has_destination(const std::string& domain,
                                  const std::string& aggregate_root) const {
    std::string key = make_key(domain, aggregate_root);
    return destinations_.find(key) != destinations_.end();
}

std::string SagaContext::make_key(const std::string& domain, const std::string& root) {
    std::ostringstream ss;
    ss << domain << ":";
    for (unsigned char c : root) {
        ss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(c);
    }
    return ss.str();
}

}  // namespace angzarr

File: client/go/saga_context.go
package angzarr

import (
	"encoding/hex"

	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
)

// SagaContext provides access to destination aggregate state for the splitter pattern.
//
// Used when one event triggers commands to multiple aggregates. Provides sequence
// number lookup for optimistic concurrency control.
//
// Example usage:
//
//	func HandleTableSettled(evt *examples.TableSettled, ctx *SagaContext) []*pb.CommandBook {
//	    commands := make([]*pb.CommandBook, 0, len(evt.Payouts))
//	    for _, payout := range evt.Payouts {
//	        seq := ctx.GetSequence("player", payout.PlayerRoot)
//	        cmd := &examples.TransferFunds{PlayerRoot: payout.PlayerRoot, Amount: payout.Amount}
//	        commands = append(commands, NewCommandBook("player", cmd, seq))
//	    }
//	    return commands
//	}
type SagaContext struct {
	destinations map[string]*pb.EventBook
}

// NewSagaContext creates a context from a list of destination EventBooks.
func NewSagaContext(destinationBooks []*pb.EventBook) *SagaContext {
	ctx := &SagaContext{
		destinations: make(map[string]*pb.EventBook),
	}
	for _, book := range destinationBooks {
		if book.GetCover() != nil && book.GetCover().GetDomain() != "" {
			key := makeSagaContextKey(book.GetCover().GetDomain(), book.GetCover().GetRoot().GetValue())
			ctx.destinations[key] = book
		}
	}
	return ctx
}

// GetSequence returns the next sequence number for a destination aggregate.
// Returns 1 if the aggregate doesn't exist yet.
func (ctx *SagaContext) GetSequence(domain string, aggregateRoot []byte) uint32 {
	key := makeSagaContextKey(domain, aggregateRoot)
	book, ok := ctx.destinations[key]
	if !ok || len(book.GetPages()) == 0 {
		return 1
	}
	lastPage := book.GetPages()[len(book.GetPages())-1]
	if header := lastPage.GetHeader(); header != nil {
		return header.GetSequence() + 1
	}
	return 1
}

// GetDestination returns the EventBook for a destination aggregate, if available.
func (ctx *SagaContext) GetDestination(domain string, aggregateRoot []byte) *pb.EventBook {
	key := makeSagaContextKey(domain, aggregateRoot)
	return ctx.destinations[key]
}

// HasDestination checks if a destination exists.
func (ctx *SagaContext) HasDestination(domain string, aggregateRoot []byte) bool {
	key := makeSagaContextKey(domain, aggregateRoot)
	_, ok := ctx.destinations[key]
	return ok
}

func makeSagaContextKey(domain string, root []byte) string {
	return domain + ":" + hex.EncodeToString(root)
}
